crates/walrus-sui/src/client/retry_client/retriable_rpc_client.rs

```diff
@@ -293,58 +301,80 @@ impl RetriableRpcClient {
+                            "success",
+                            start_time.elapsed(),
+                        );
+                    }
+                    self.reset_fullnode_failure_metrics();
+                    return Ok(checkpoint);
+                }
+                Err(error) => error,
+            };
+
+            self.num_failures.fetch_add(1, Ordering::Relaxed);
+            tracing::debug!(?error, "primary client error while fetching checkpoint");
+
+            // If the fallback client is not set, or the error is not eligible for fallback,
+            if self.fallback_client.is_none()
+                || !self
+                    .fallback_client
+                    .as_ref()
+                    .unwrap()
+                    .is_eligible_for_fallback(
+                        sequence_number,
+                        &error,
+                        self.last_success.load(Ordering::Relaxed),
+                        self.num_failures.load(Ordering::Relaxed),
+                    )
+            {
+                tracing::debug!(
+                    failback_client_set = ?self.fallback_client.is_some(),
+                    "primary client error while fetching checkpoint is not eligible for fallback"
+                );
                 if let Some(metrics) = self.metrics.as_ref() {
                     metrics.record_rpc_latency(
                         "get_full_checkpoint",
                         &self.client.get_current_rpc_url().await,
-                        "success",
+                        "failure",
                         start_time.elapsed(),
                     );
                 }
-                self.reset_fullnode_failure_metrics();
-                return Ok(checkpoint);
-            }
-            Err(error) => error,
-        };
-
-        self.num_failures.fetch_add(1, Ordering::Relaxed);
-        tracing::debug!(?error, "primary client error while fetching checkpoint");
-        let Some(ref fallback) = self.fallback_client else {
-            if let Some(metrics) = self.metrics.as_ref() {
-                metrics.record_rpc_latency(
-                    "get_full_checkpoint",
-                    &self.client.get_current_rpc_url().await,
-                    "failure",
-                    start_time.elapsed(),
-                )
+                return Err(error);
             }
-            return Err(error);
-        };
 
-        if !error.is_eligible_for_fallback(
-            sequence_number,
-            self.last_success.load(Ordering::Relaxed),
-            self.num_failures.load(Ordering::Relaxed),
-        ) {
-            tracing::debug!(
-                "primary client error while fetching checkpoint is not eligible for fallback"
+            // RPC node failure is sustained, so we skip it for 5 minutes.
+            // Also, at this point, the fallback client must be set.
+            self.skip_rpc_for_checkpoint_until.store(
```

Comment:
Does this also get set for something like missing events error? Since it is in the check for `is_elligible_for_fallback`

Reply:
Great catch! We should only skip calling RPC node when `is_failure_window_exceeded`. (one thing I don't feel to comfortable about this PR is that, for realistic scenario like this, it's very hard to test/discover such corner cases).